Add ability to display torrent "privateness" in UI

ManiMatter opened the pull request:
Adds private flag to GUI that is available in /info and /properties response

+ WebUI & GUI: +"Private" column in torrent list (hidden by default)
+ WebUI & GUI: +"Private" flag in properties of selected torrent ("Information" section)

ManiMatter commented:
oncrement API_VERSION in webapplication.h due to changes to private flag

ManiMatter commented:
@Chocobo1 , @glassez - I would appreciate your review. 

I tested it and it works. Only one small item that is not yet working as expected, maybe you can point me in the right direction? On the GUI, sorting by the "Private" column does not work. On the WebGUI, it works as expected. 

My hunch is that I am making some mistake around here, but I have not been able to figure it out.
https://github.com/qbittorrent/qBittorrent/blob/b5120765185b6363008bb36a386ddbb4866d7705/src/gui/transferlistsortmodel.cpp#L170-L171

thalieht commented on src/gui/transferlistmodel.cpp:
```diff
@@ -193,6 +193,7 @@ QVariant TransferListModel::headerData(const int section, const Qt::Orientation
             case TR_INFOHASH_V1: return tr("Info Hash v1", "i.e: torrent info hash v1");
             case TR_INFOHASH_V2: return tr("Info Hash v2", "i.e: torrent info hash v2");
             case TR_REANNOUNCE: return tr("Reannounce In", "Indicates the time until next trackers reannounce");
+            case TR_PRIVATE: return tr("Private", "Flags torrents form private trackers");
```
Whether a torrent is private isn't defined by its source. Is disambiguation needed for this? I don't have any suggestions.

ManiMatter replied:
How about simply putting "Flags private torrents"?

thalieht replied:
Sure.

ManiMatter replied:
Done

thalieht commented on src/gui/transferlistmodel.cpp:
```diff
@@ -357,6 +358,13 @@ QString TransferListModel::displayValue(const BitTorrent::Torrent *torrent, cons
+    {
+        if (isPrivate)
+            return QStringLiteral("Yes");
+        return QStringLiteral("No");
```
```suggestion
        return tr("No");
```

Same in all other places where you use `QStringLiteral`

ManiMatter replied:
Done

thalieht commented on src/gui/transferlistmodel.cpp:
```diff
@@ -578,6 +588,7 @@ QVariant TransferListModel::data(const QModelIndex &index, const int role) const
         case TR_LAST_ACTIVITY:
         case TR_AVAILABILITY:
         case TR_REANNOUNCE:
+        case TR_PRIVATE:
```
```suggestion
```
Strings are left aligned.

thalieht commented:
>On the GUI, sorting by the "Private" column does not work.

You have to add the column to `TransferlistModel::InternalValue`

Problem in General tab in GUI only:
- Select a private torrent
- Select a torrent without metadata
- Private field in General tab is not updated to "No"

ManiMatter commented:
Thanks for the review and pointers, @thalieht 

> > On the GUI, sorting by the "Private" column does not work.
> 
> You have to add the column to `TransferlistModel::InternalValue`

Spot on. Now it works.

> Problem in General tab in GUI only:
> 
> * Select a private torrent
> * Select a torrent without metadata
> * Private field in General tab is not updated to "No"

I am trying to reproduce this. What's the best way to generate a torrent without metadata?

thalieht commented:
>What's the best way to generate a torrent without metadata?

Add a magnet link to session and pause it before its metadata are retrieved.

Chocobo1 commented on src/gui/transferlistmodel.cpp:
```diff
@@ -357,6 +358,13 @@ QString TransferListModel::displayValue(const BitTorrent::Torrent *torrent, cons
         return Utils::Misc::userFriendlyDuration(time);
     };
 
+    const auto privateString = [](const bool isPrivate) -> QString
+    {
+        if (isPrivate)
+            return tr("Yes");
+        return tr("No");
+    };
```
Maybe add `hideValues`?
```suggestion
    const auto privateString = [hideValues](const bool isPrivate) -> QString
    {
        if (hideValues && !isPrivate)
            return {};
        return isPrivate ? : tr("Yes") : tr("No");
    };
```

Chocobo1 commented on src/gui/properties/propertieswidget.cpp:
```diff
@@ -309,6 +309,10 @@ void PropertiesWidget::loadTorrentInfos(BitTorrent::Torrent *const torrent)
         m_ui->labelCommentVal->setText(Utils::Misc::parseHtmlLinks(m_torrent->comment().toHtmlEscaped()));
 
         m_ui->labelCreatedByVal->setText(m_torrent->creator());
+
+        // Private
+        m_ui->labelPrivateVal->setText((m_torrent && m_torrent->isPrivate()) ? tr("Yes") : tr("No"));
```
Does it depend on metadata? If not, it should be moved out (above) of the if conditional.

Chocobo1 replied:
This check is redundant anyway:
```suggestion

        // Private
        m_ui->labelPrivateVal->setText(m_torrent->isPrivate() ? tr("Yes") : tr("No"));
```

glassez replied:
> Does it depend on metadata? If not, it should be moved out (above) of the if conditional.

AFAIK, we can't know whether torrent is private or not if there is no metadata available. So I would display "N/A" in case when there are no metadata available.

Chocobo1 commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -167,6 +167,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
     case TransferListModel::TR_NAME:
     case TransferListModel::TR_SAVE_PATH:
     case TransferListModel::TR_TRACKER:
+    case TransferListModel::TR_PRIVATE:
```
Keep it sorted:
```suggestion
    case TransferListModel::TR_PRIVATE:
    case TransferListModel::TR_SAVE_PATH:
    case TransferListModel::TR_TRACKER:
```

Chocobo1 commented on src/webui/www/private/scripts/prop-general.js:
```diff
@@ -211,6 +212,11 @@ window.qBittorrent.PropGeneral = (function() {
 
                     $("comment").set("html", window.qBittorrent.Misc.parseHtmlLinks(window.qBittorrent.Misc.escapeHtml(data.comment)));
+
+                    const is_private = (data.is_private)
+                        ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]"
+                        : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]";
+                    $("private").set("html", is_private);
```
```suggestion
                    const isPrivate = data.private
                        ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]"
                        : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]";
                    $("private").set("html", isPrivate);
```

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1332,6 +1333,14 @@ window.qBittorrent.DynamicTable = (function() {
+
+            // private
+            this.columns["is_private"].updateTd = function(td, row) {
+                const is_private = this.getRowValue(row);
+                const string = (is_private === true) ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]" : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]";
```
```suggestion
                const isPrivate = (this.getRowValue(row) === true);
                const string = isPrivate ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]" : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]";
```

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -949,6 +949,7 @@ window.qBittorrent.DynamicTable = (function() {
             this.newColumn("last_activity", "", "QBT_TR(Last Activity)QBT_TR[CONTEXT=TransferListModel]", 100, false);
             this.newColumn("availability", "", "QBT_TR(Availability)QBT_TR[CONTEXT=TransferListModel]", 100, false);
             this.newColumn("reannounce", "", "QBT_TR(Reannounce In)QBT_TR[CONTEXT=TransferListModel]", 100, false);
+            this.newColumn("is_private", "", "QBT_TR(Private)QBT_TR[CONTEXT=TransferListModel]", 100, false);
```
Try:
```suggestion
            this.newColumn("private", "", "QBT_TR(Private)QBT_TR[CONTEXT=TransferListModel]", 100, false);
```

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1332,6 +1333,14 @@ window.qBittorrent.DynamicTable = (function() {
             };
+
+            // private
+            this.columns["is_private"].updateTd = function(td, row) {
```
```suggestion
            this.columns["private"].updateTd = (td, row) => {
```

ManiMatter commented:
> > What's the best way to generate a torrent without metadata?
> 
> Add a magnet link to session and pause it before its metadata are retrieved.

Thank you, I've been able to reproduce it. I'm a bit lost as to why this happens.
I tried validating the "m_torrent" before the isPrivate() like this, but it didn't do the trick:
https://github.com/qbittorrent/qBittorrent/blob/42a56f62e220bcc465092cedada5db6e6b95e249/src/gui/properties/propertieswidget.cpp#L314

Any ideas?

Chocobo1 commented on src/gui/properties/propertieswidget.ui:
```diff
@@ -871,6 +871,22 @@
                </widget>
               </item>
               <item row="4" column="0">
+               <widget class="QLabel" name="labelPrivate">
```
IMO it should be placed above "Info Hash v1".
Like so:
```
...
Added On: | Completed On: | Created On:
Private:
Info Hash v1:
...
```

stalkerok replied:
Maybe better 
```
...
Added On: | Completed On: | Created On:
Info Hash v1: | Private:
...
```
it'll save space.

ManiMatter replied:
Personally, I prefer @Chocobo1 order. If space is the issue, I would find it more logical to put the two hashes on to the same line, rather than Hash v1 and Private.

stalkerok replied:
@ManiMatter I would still prefer to keep the hash to two lines.
@thalieht @glassez do you have your own preferences?

ManiMatter replied:
@stalkerok - I moved the private flag up as per @Chocobo1 and left the hashes on two lines for now. 
Screenshots of current implementation: https://github.com/qbittorrent/qBittorrent/pull/20951#issuecomment-2184175736

thalieht replied:
I don't have an opinion on this.

stalkerok replied:
@ManiMatter, however, I would still put it on one line. Or so, if possible, then the hash would definitely not overlap
```
...
Added On: | Completed On: | Created On:
Info Hash v1: | <empty> | Private:
...
```

stalkerok replied:
Anyway, it's not that important, maybe some more information will take up the empty space in the future.

Chocobo1 commented on src/webui/www/private/views/properties.html:
```diff
@@ -82,6 +82,10 @@
                 <td class="generalLabel">QBT_TR(Info Hash v2:)QBT_TR[CONTEXT=PropertiesWidget]</td>
                 <td colspan="5" id="torrent_hash_v2"></td>
             </tr>
+            <tr>
+                <td class="generalLabel">QBT_TR(Private:)QBT_TR[CONTEXT=PropertiesWidget]</td>
+                <td colspan="5" id="private"></td>
+            </tr>
```
This should be moved up and drop the `colspan`:
```suggestion
            <tr>
                <td class="generalLabel">QBT_TR(Private:)QBT_TR[CONTEXT=PropertiesWidget]</td>
                <td id="private"></td>
            </tr>
```

stalkerok commented:
> > What's the best way to generate a torrent without metadata?
> 
> Add a magnet link to session and pause it before its metadata are retrieved.

This can also be done by adding a non-existent magnet link or you can use an existing one by changing a few characters.

ManiMatter commented:
> Problem in General tab in GUI only:
> 
> Select a private torrent
> Select a torrent without metadata
> Private field in General tab is not updated to "No"

Does anybody have an idea how to solve it?

ManiMatter commented:
> > Problem in General tab in GUI only:
> > Select a private torrent
> > Select a torrent without metadata
> > Private field in General tab is not updated to "No"
> 
> Does anybody have an idea how to solve it?

Anyone? Already a pointer where to look to find a solution would be appreciated. I‘m tapping in the dark as to why this happens.. 

In the WebAPI I see that the private status for torrents without metadata is false. However, switching in the gui (bot the Webgui) from a private torrent to a torrent without metadata keeps the flag on ‚true‘, which makes me believe it is not false but null or something similar in the torrent data model, for which the update fails. 

But I‘m lost as to how & where to tackle this. 

Thoughts anyone? Thanks for any pointers.

thalieht commented:
>Anyone? Already a pointer where to look to find a solution would be appreciated. I‘m tapping in the dark as to why this happens..

This will take care of it: https://github.com/qbittorrent/qBittorrent/pull/20951#discussion_r1640832571

ManiMatter commented:
> > Anyone? Already a pointer where to look to find a solution would be appreciated. I‘m tapping in the dark as to why this happens..
> 
> This will take care of it: [#20951 (comment)](https://github.com/qbittorrent/qBittorrent/pull/20951#discussion_r1640832571)

Hi @thalieht 
This change reverts to the state as I had it originally; the redundant torrent check I had added as an attempt to fix the problem you raised (without success). I reverted it (as @Chocobo1 suggested), but now of course we are back to step 1 where selecting a torrent without metadata will flag "private: yes" if selected after a private torrent in the GUI 

Any pointers still appreciated how to solve this; apologies you already have pointed me in the right direction and I have not picked it up

thalieht commented:
Something like this, (not sure if multiple conditional operators in the same line are allowed), also you missed this https://github.com/qbittorrent/qBittorrent/pull/20951#discussion_r1640867755 (added below)
```
--- a/src/gui/properties/propertieswidget.cpp
+++ b/src/gui/properties/propertieswidget.cpp
@@ -324,7 +324,10 @@ void PropertiesWidget::loadTorrentInfos(BitTorrent::Torrent
 *const torrent)
     m_ui->labelInfohash2Val->setText(m_torrent->infoHash().v2().isValid() ? m_t
orrent->infoHash().v2().toString() : tr("N/A"));
     // URL seeds
     loadUrlSeeds();
-    if (m_torrent->hasMetadata())
+    const bool hasMetadata = m_torrent->hasMetadata();
+    // Private
+    m_ui->labelPrivateVal->setText(m_torrent->isPrivate() ? tr("Yes") : !hasMetadata ? tr("N/A") : tr("No"));
+    if (hasMetadata)
     {
         // Creation date
         m_ui->labelCreatedOnVal->setText(QLocale().toString(m_torrent->creation
Date(), QLocale::ShortFormat));
@@ -335,10 +338,6 @@ void PropertiesWidget::loadTorrentInfos(BitTorrent::Torrent
 *const torrent)
         m_ui->labelCommentVal->setText(Utils::Misc::parseHtmlLinks(m_torrent->comment().toHtmlEscaped()));
 
         m_ui->labelCreatedByVal->setText(m_torrent->creator());
-
-        // Private
-        m_ui->labelPrivateVal->setText(m_torrent->isPrivate() ? tr("Yes") : tr("No"));
-
     }
     // Load dynamic data
     loadDynamicData();
```

glassez commented:
@thalieht
I would rewrite it in slightly different way:
```c++
if (m_torrent->hasMetadata())
{
    // existing code

    m_ui->labelPrivateVal->setText(m_torrent->isPrivate() ? tr("Yes") : tr("No"));
}
else
{
    m_ui->labelPrivateVal->setText(tr("N/A"));
}
```

glassez commented:
> Adds private flag to GUI that is available in /info and /properties response
> 
> * WebUI & GUI: +"Private" column in torrent list (hidden by default)
> * WebUI & GUI: +"Private" flag in properties of selected torrent ("Information" section)

It would be nice to see screenshots.

ManiMatter commented:
@glassez @thalieht 

thanks for the inputs. Will have a look how to add this in the GUI/WebUI both for the table and the properties section. 
For the api responses of the /info and the /properties endpoint, the value returned for items that don't have metadata is still "false". I feel here we probably should not return the "private" flag at all, so that the absence of it is consistent with "N/A" behavior on frontend, and "false" means it's properly false. 

What do you think?

glassez commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -948,6 +948,7 @@ window.qBittorrent.DynamicTable = (function() {
             this.newColumn("last_activity", "", "QBT_TR(Last Activity)QBT_TR[CONTEXT=TransferListModel]", 100, false);
             this.newColumn("availability", "", "QBT_TR(Availability)QBT_TR[CONTEXT=TransferListModel]", 100, false);
             this.newColumn("reannounce", "", "QBT_TR(Reannounce In)QBT_TR[CONTEXT=TransferListModel]", 100, false);
+            this.newColumn("is_private", "", "QBT_TR(Private)QBT_TR[CONTEXT=TransferListModel]", 100, false);
```
Shouldn't it be "private" instead of "is_private"? Or is it not the binding with the data returned by WebAPI?

ManiMatter commented:
> > Adds private flag to GUI that is available in /info and /properties response
> > 
> > * WebUI & GUI: +"Private" column in torrent list (hidden by default)
> > * WebUI & GUI: +"Private" flag in properties of selected torrent ("Information" section)
> 
> It would be nice to see screenshots.

### GUI:
<img width="691" alt="image" src="https://github.com/qbittorrent/qBittorrent/assets/124743318/49a94c10-6df6-47f6-b437-ab0ed06c5e58">

### WebGUI:
<img width="950" alt="image" src="https://github.com/qbittorrent/qBittorrent/assets/124743318/273556e8-29b0-4794-af32-fd8bd99dc52f">

ManiMatter commented:
Here's a summary of what I still see as open:
1) The `/properties` and the `/info` API endpoints return `private: false` for torrents without metadata. My view: "private" flag should not even be included. My hunch: That needs to be changed in the `serialize_torrent.cpp`. Do you agree?

2) Torrents without metadata return "No" in the WebUI in the table as well as the properties section. My hunch: Once above (1) is changed, the logic that I have already added should work. Let me know if you see it differently.

3) Torrents without metadata return "No" in the GUI in the table as well as the properties section. Here, I followed your snippet above, but for some reason, it's ignored. Any pointers appreciated.

4) The bug that does not reset the private flag in the properties section of the GUI for torrents without metadata is still there. My hunch is that once (3) is resolved, this will be solved, too. Let me know if you see it differently.

thalieht commented on src/gui/transferlistmodel.cpp:
```diff
@@ -431,6 +439,8 @@ QString TransferListModel::displayValue(const BitTorrent::Torrent *torrent, cons
     case TR_REANNOUNCE:
         return reannounceString(torrent->nextAnnounce());
+    case TR_PRIVATE:
+        return privateString(torrent->isPrivate());
```
With the changes in the lambda above:
```suggestion
        return privateString(torrent->isPrivate(), torrent->hasMetadata());
```

thalieht commented on src/gui/properties/propertieswidget.cpp:
```diff
@@ -335,6 +335,18 @@ void PropertiesWidget::loadTorrentInfos(BitTorrent::Torrent *const torrent)
         m_ui->labelCreatedByVal->setText(m_torrent->creator());
+
+        // Private
+        if (m_torrent->hasMetadata())
```
>4. The bug that does not reset the private flag in the properties section of the GUI for torrents without metadata is still there.

This already exists in line 327 (at the moment of writing). https://github.com/qbittorrent/qBittorrent/pull/20951#issuecomment-2184100600 was referring to it.

/off topic
I can't help with API stuff.

ManiMatter replied:
Thank you. Fixed it.

glassez commented on src/gui/transferlistmodel.cpp:
```diff
@@ -357,6 +358,13 @@ QString TransferListModel::displayValue(const BitTorrent::Torrent *torrent, cons
+        if (hideValues && !isPrivate)
+            return {};
+        return isPrivate ? tr("Yes") : tr("No");
+    };
```
I believe it should display only "Yes" values when `hideValues` is `true`.

glassez replied:
Another idea. Should we always display an empty string in "Private" column for torrents w/o metadata (instead of "N/A")?

glassez commented:
> 1. My hunch: That needs to be changed in the `serialize_torrent.cpp`. Do you agree?

No.
This serialization is also used by synchronization API support routines, which expects it to produce structures of the same layout (in order to be compared field by field). FYI, WebUI torrents table uses "sync" API, not "torrents/info".
The solution could be to return a "null" value if there is no metadata, e.g.:
```c++
 {KEY_TORRENT_PRIVATE, (torrent.hasMetadata() ? torrent.isPrivate() : QVariant()}),
```
But I think it would be easier to go the same way that is used in the main application code, i.e. provide "hasMetadata" field via WebAPI and use it in WebUI.

ManiMatter commented:
> > 1. My hunch: That needs to be changed in the `serialize_torrent.cpp`. Do you agree?
> 
> No. This serialization is also used by synchronization API support routines, which expects it to produce structures of the same layout (in order to be compared field by field). FYI, WebUI torrents table uses "sync" API, not "torrents/info". The solution could be to return a "null" value if there is no metadata, e.g.:
> 
> ```c++
>  {KEY_TORRENT_PRIVATE, (torrent.hasMetadata() ? torrent.isPrivate() : QVariant()}),
> ```
> 
> But I think it would be easier to go the same way that is used in the main application code, i.e. provide "hasMetadata" field via WebAPI and use it in WebUI.

I think both make sense. I'd expect "null" to be returned for KEY_TORRENT_PRIVATE if it's neither true nor false (since no metadata), and that we return the hasMetadata as well, and then use this info in the WebUI

ManiMatter commented:
sorry - I somehow messed up my branch, which closed this PR. Apologies for this. Will continue with integrating your feedback, and then create a new PR. Apologies

ManiMatter commented:
@glassez 
I have implemented [your suggestion](https://github.com/qbittorrent/qBittorrent/pull/20951#issuecomment-2184938138) for the /info API, and am struggling to do the same on the /properties API.

What works for now is that in case there is no metadata, I return "-1" for the "private" key, but in reality I think we should be returning "null" rather than "-1". Replacing "-1" with "QVariant()" seems not to work. Any ideas?


https://github.com/qbittorrent/qBittorrent/blob/5a0cc031f6c5d02d9523f9fc4ec5ac59dcc27463/src/webui/api/torrentscontroller.cpp#L481

glassez commented:
> Replacing "-1" with "QVariant()" seems not to work.

Sure. After all, `QJsonObject` is being constructed here, not `QVariantMap`. So you should use `QJsonValue()` instead of `QVariant()`.

stalkerok commented:
CI - Windows won't start, crashes.

ManiMatter commented:
> CI - Windows won't start, crashes.

Thanks for spotting this. I unfortunately have no windows experience. Any ideas why it happens?

Chocobo1 commented on src/gui/properties/propertieswidget.ui:
```diff
@@ -823,39 +823,58 @@
                 </property>
                 <property name="alignment">
-                 <set>Qt::AlignRight|Qt::AlignTrailing|Qt::AlignVCenter</set>
+                 <set>Qt::AlignRight|Qt::AlignTop|Qt::AlignTrailing</set>
```
```suggestion
                 <set>Qt::AlignRight|Qt::AlignTrailing|Qt::AlignVCenter</set>
```

Chocobo1 commented on src/gui/properties/propertieswidget.ui:
```diff
@@ -823,39 +823,58 @@
+                <property name="textFormat">
+                 <enum>Qt::PlainText</enum>
+                </property>
+                <property name="wordWrap">
+                 <bool>true</bool>
+                </property>
```
```suggestion
```

ManiMatter commented on src/webui/api/serialize/serialize_torrent.h:
```diff
@@ -94,5 +94,6 @@ inline const QString KEY_TORRENT_AVAILABILITY = u"availability"_s;
 inline const QString KEY_TORRENT_REANNOUNCE = u"reannounce"_s;
 inline const QString KEY_TORRENT_COMMENT = u"comment"_s;
 inline const QString KEY_TORRENT_PRIVATE = u"private"_s;
+inline const QString KEY_TORRENT_HAS_METADATA = u"has_metadata"_s;
```
Why "have" and not "has"? While "metadata" is plural, it's the torrent (singular) that has (or does not have) metadata. Thus "has" to me sounds grammatically correct?

Chocobo1 replied:
From https://preply.com/en/blog/has-vs-have/ :
>The main difference between “has” and “have” lies in which subjects they use. “Has” is used with the third-person singular subjects “he,” “she,” and “it,” or a singular noun. “Have” is used with first-person (“I,” “we”), second-person (“you”), and third-person plural (“they”) subjects.

The "metadata" is not a 'third-person' property/subject but a native property of a `torrent` object.

ManiMatter replied:
I understand that metadata is a native property, but in my view we describe "the torrent", which is the third person singular. It ("the torrent") has properties. 

If the torrent described itself, it would say "I have" properties. But here we are the people describing, thus I'd argue we should say "It has", and in consequence stick with "hasMetadata". It would also be consistent with the torrent data model which uses "hasMetadata()"

What do the others think? @thalieht @stalkerok  @glassez

Chocobo1 replied:
Hmmm after some googling [.NET naming guidelines](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members?redirectedfrom=MSDN) uses "has":
>Optionally, you can also prefix Boolean properties with "Is", "Can", or "Has", but only where it adds value.

Well. It seems my opinion is not conclusive. Please disregard it.

glassez replied:
> Hmmm after some googling [.NET naming guidelines](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members?redirectedfrom=MSDN) uses "has":
> 
> > Optionally, you can also prefix Boolean properties with "Is", "Can", or "Has", but only where it adds value.
> 
> Well. It seems my opinion is not conclusive. Please disregard it. Sorry for the noise.

I believe that such naming rules based on ordinary wording, for example, "torrent has metadata".
But Qt naming convention is more abstract from natural language. They require not to use third from for verbs (`is` prefix is special case):
>Verbs have no prefix and don't use the third person (-s)

I suggest not to dwell on it here, since we don't have clear rules about it.

ManiMatter commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -165,6 +165,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
     case TransferListModel::TR_CATEGORY:
     case TransferListModel::TR_DOWNLOAD_PATH:
     case TransferListModel::TR_NAME:
+    case TransferListModel::TR_PRIVATE:
```
Will do. For my education so I get it right in the future; what's the difference?

ManiMatter commented on src/webui/www/private/scripts/prop-general.js:
```diff
@@ -210,6 +211,13 @@ window.qBittorrent.PropGeneral = (function() {
+                        : (data.private === true
+                            ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]"
+                            : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]");
+                    $("private").set("html", isPrivate);
```
I think there is something off with this change; the linter complains about a curly bracket after the merge. Do you spot it? Else I will check it when I find the time

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1331,6 +1332,20 @@ window.qBittorrent.DynamicTable = (function() {
+                    : "QBT_TR(N/A)QBT_TR[CONTEXT=PropertiesWidget]";
+                td.set("text", string);
+                td.set("title", string);
+            };
+
```
```suggestion
            };
```

Chocobo1 replied:
```suggestion
```

Chocobo1 commented on src/webui/www/private/scripts/prop-general.js:
```diff
@@ -210,6 +211,13 @@ window.qBittorrent.PropGeneral = (function() {
 
                     $("comment").set("html", window.qBittorrent.Misc.parseHtmlLinks(window.qBittorrent.Misc.escapeHtml(data.comment)));
+
+                    if (data.has_metadata) {
+                        $("private").set("html", (data.private ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]" : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]"));
+                    }
+                    else {
+                        $("private").set("html", "QBT_TR(N/A)QBT_TR[CONTEXT=PropertiesWidget]");
+                    }
```
This should address the CI error:
```suggestion
                    if (data.has_metadata) {
                        $("private").set("html", (data.private
                            ? "QBT_TR(Yes)QBT_TR[CONTEXT=PropertiesWidget]"
                            : "QBT_TR(No)QBT_TR[CONTEXT=PropertiesWidget]"));
                    }
                    else {
                        $("private").set("html", "QBT_TR(N/A)QBT_TR[CONTEXT=PropertiesWidget]");
                    }
```

ManiMatter replied:
Thx, that passes now

stalkerok commented:
<details>
  <summary>Event Log</summary>
Имя сбойного приложения: qbittorrent.exe, версия: 5.0.0.0, метка времени: 0x667abeea

Имя сбойного модуля: MSVCP140.dll, версия: 14.36.32532.0, метка времени: 0x04a30cf0
Код исключения: 0xc0000005
Смещение ошибки: 0x0000000000012f58
Идентификатор сбойного процесса: 0x179c
Время запуска сбойного приложения: 0x01dac732e39f5c7d
Путь сбойного приложения: ...\qbittorrent test\5948 Add ability to display torrent privateness in UI\qBittorrent\qbittorrent.exe
Путь сбойного модуля: C:\Windows\SYSTEM32\MSVCP140.dll
Идентификатор отчета: f2f90df4-6265-45c2-8b09-6b9c79f1d527
Полное имя сбойного пакета: 
Код приложения, связанного со сбойным пакетом: 
</details>

[qbittorrent.exe.6044.dmp](https://github.com/user-attachments/files/15976999/qbittorrent.exe.6044.dmp)

ManiMatter commented:
Thanks @stalkerok 
I just opened it on mac visual studio and it looks like something that can‘t be read, ie lots of gibberish characters. 

I think we need some help here from somebody who knows Windows…

stalkerok commented:
Perhaps the language of my system plays a role in this.

ManiMatter commented:
> Perhaps the language of my system plays a role in this.

Based on the logs above I suppose that is Russian/Cyrillic?

Is the crash also happening with the windows CI of the latest merge to the main branch or only in this feature branch? Trying to understand whether it‘s a general problem or specific to the changes here.

xavier2k6 commented:
```
Exception Information
MSVCP140!MTX_DO_LOCK+9CIn qbittorrent.exe.6044.dmp the assembly instruction at msvcp140!mtx_do_lock+9c in C:\Windows\System32\msvcp140.dll from Microsoft Corporation has caused an access violation exception (0xC0000005) when trying to read from memory location 0x0000[0](https://github.com/qbittorrent/qBittorrent/pull/20951#B1C1Thread3104)000 on thread 0
```

@stalkerok Can you grab/install latest Microsoft Visual C++ Redistributable & see if crash still happens.
[Microsoft Visual C++ Redistributable Latest Supported Downloads](https://docs.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022)

stalkerok commented:
> Can you grab/install latest Microsoft Visual C++ Redistributable & see if crash still happens.

Great, there was no crash.

thalieht commented:
>Can you grab/install latest Microsoft Visual C++ Redistributable & see if crash still happens.

Turns out i didn't have the latest... all good now, thanks!

stalkerok commented:
If this will affect running the 5.0 release version, it is worth warning users to install the latest MSVC libraries.

xavier2k6 commented:
> If this will affect running the 5.0 release version, it is worth warning users to install the latest MSVC libraries.

Requiring latest runtime only happens to our CI builds every once in a while.

ManiMatter commented:
Shall we tag this to milestone 5.0?

glassez commented:
> Shall we tag this to milestone 5.0?

I don't think this is a good idea for a PR that doesn't have a predictable completion date. It may not make it to v5.0 in time.

ManiMatter commented:
@glassez - I was under the impression that this one is pretty much done. Do you see remaining items that need to be addressed?

glassez commented:
> @glassez - I was under the impression that this one is pretty much done. Do you see remaining items that need to be addressed?

Yes. At least https://github.com/qbittorrent/qBittorrent/pull/20951#discussion_r1657005692.

ManiMatter commented on src/gui/transferlistmodel.cpp:
```diff
@@ -512,6 +525,8 @@ QVariant TransferListModel::internalValue(const BitTorrent::Torrent *torrent, co
         return QVariant::fromValue(torrent->infoHash().v2());
     case TR_REANNOUNCE:
         return torrent->nextAnnounce();
+    case TR_PRIVATE:
```
Thanks - would this work?https://github.com/qbittorrent/qBittorrent/pull/20951/commits/4595dda2ef8be3d99033e91f17e8144b56d428ec

glassez replied:
> Thanks - would this work?[4595dda](https://github.com/qbittorrent/qBittorrent/commit/4595dda2ef8be3d99033e91f17e8144b56d428ec)

No.
Did I suggest returning its string representation?
Isn't it obvious that `internalValue()` should provide something else (except, of course, the cases when the data itself is string)?
IMO, I made it clear that it is only necessary to add some third value that would correspond to the case when the "privateness" is not available (due to the lack of metadata).

ManiMatter replied:
I thought I'd keep the code consistent with with displayValue(); Clearly, I misinterpreted your suggestion. 

I amended the code. Is this how you had it in mind? https://github.com/qbittorrent/qBittorrent/pull/20951/commits/d60c77ca6ab57e9f973db67ad4b55a14598b3827

glassez commented on src/gui/properties/propertieswidget.cpp:
```diff
@@ -335,7 +336,16 @@ void PropertiesWidget::loadTorrentInfos(BitTorrent::Torrent *const torrent)
 
         m_ui->labelCreatedByVal->setText(m_torrent->creator());
+
+        // Private
```
```suggestion
```

glassez commented on src/gui/properties/propertieswidget.cpp:
```diff
@@ -335,7 +336,16 @@ void PropertiesWidget::loadTorrentInfos(BitTorrent::Torrent *const torrent)
     }
+    else
+    {
+        // Private (when there is no metadata)
```
```suggestion
```

glassez replied:
It makes no sense to write obvious things in the comments, for example, **what** the code does (this is already clear from the code itself). If you want to explain it, then write down **why** it does it.

glassez replied:
P.S. Below is an example of a completely meaningless comment.
```c++
// Load dynamic data
loadDynamicData();
```

ManiMatter replied:
Thanks... I agree with you that the comment is superfluous. I only added it because the code that is already there has the same type of comments.. 

Example: https://github.com/qbittorrent/qBittorrent/blob/d60c77ca6ab57e9f973db67ad4b55a14598b3827/src/gui/properties/propertieswidget.cpp#L321-L339

glassez commented on src/gui/transferlistmodel.cpp:
```diff
@@ -512,6 +531,8 @@ QVariant TransferListModel::internalValue(const BitTorrent::Torrent *torrent, co
     case TR_REANNOUNCE:
         return torrent->nextAnnounce();
+    case TR_PRIVATE:
+        return privateSort(torrent->isPrivate(), torrent->hasMetadata());
```
Why to complicate it?
```suggestion
        return (torrent->hasMetadata() ? torrent->isPrivate() : -1);
```

glassez replied:
Seems more correct would be the following:
```c++
return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());
```
...and use appropriate comparison in sort model (that takes into account invalid values).

ManiMatter replied:
> Seems more correct would be the following:
> 
> ```c++
> return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());
> ```
> 
> ...and use appropriate comparison in sort model (that takes into account invalid values).

I implemented it:
https://github.com/qbittorrent/qBittorrent/blob/f565c33a1576de8f62a50cef2b254a7a58555052/src/gui/transferlistmodel.cpp#L528

glassez replied:
> I implemented it:
> 
> https://github.com/qbittorrent/qBittorrent/blob/f565c33a1576de8f62a50cef2b254a7a58555052/src/gui/transferlistmodel.cpp#L528

OK.
But you still haven't taken care of the second part:
>...and use appropriate comparison in sort model (that takes into account invalid values).

IIRC, the following will not handle invalid values properly:
```c++
threeWayCompare(leftValue.toInt(), rightValue.toInt());
```

ManiMatter replied:
I am still not sure why we changed 
approach a)
> return (torrent->hasMetadata() ? torrent->isPrivate() : -1);

to approach b)
> return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());

With the first one, we always get an integer (1, 0, -1), since -1 is the fall-back and isPrivate() is only 1 or 0.
Since we convert the arguments for the threewaycompare explicitly to int, shouldn't it work if we sticked with approach a)?
https://github.com/qbittorrent/qBittorrent/blob/f565c33a1576de8f62a50cef2b254a7a58555052/src/gui/transferlistsortmodel.cpp#L204

glassez replied:
> I am still not sure why we changed approach a)
> 
> > return (torrent->hasMetadata() ? torrent->isPrivate() : -1);
> 
> to approach b)
> 
> > return (torrent->hasMetadata() ? torrent->isPrivate() : QVariant());

Because it better (more explicitly) reflects its essence. It either returns the value of "privateness "itself (of type bool), or invalid if it is not available (via an uninitialized QVariant). Another way is a workaround, since all values (both valid and not) are represented by another data type (integer). I.e. true/false/invalid vs 1/0/-1.

glassez replied:
> Since we convert the arguments for the threewaycompare explicitly to int

`internalValue()` should just return internal/raw representation of data. It shouldn't really depend on some particular usage of it.

ManiMatter replied:
Thank you for the explanation, @glassez

glassez commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -199,6 +199,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
     case TransferListModel::TR_POPULARITY:
         return customCompare(leftValue.toReal(), rightValue.toReal());
 
+    case TransferListModel::TR_PRIVATE:
```
Try to add the following function:
```c++
    int customCompare(const QVariant &left, const QVariant &right)
    {
        const bool isLeftValid = left.isValid();
        const bool isRightValid = right.isValid();

        if (isLeftValid == isRightValid)
            return threeWayCompare(left, right);
        return isLeftValid ? -1 : 1;
    }
``` 

...and use it to compare "privateness":
```c++
case TransferListModel::TR_PRIVATE:
    return customCompare(leftValue, rightValue);
```

glassez replied:
> Try to add the following function:
> 
> ```c++
>     int customCompare(const QVariant &left, const QVariant &right)
>     {
>         const bool isLeftValid = left.isValid();
>         const bool isRightValid = right.isValid();
> 
>         if (isLeftValid == isRightValid)
>             return threeWayCompare(left, right);
>         return isLeftValid ? -1 : 1;
>     }
> ```

Or maybe convert an existing one into a template:
```c++
// this
int customCompare(const QDateTime &left, const QDateTime &right)
// into
template <typename T>
int customCompare(const T &left, const T &right)
```

ManiMatter replied:
Please forgive me for maybe being slow in understanding.

In my latest commit, I simply moved the `case TransferListModel::TR_PRIVATE:` from a `threeWayCompare` to a `customCompare`. 

However, I did not convert an existing function into a template, because I believe that is already existing?:

https://github.com/qbittorrent/qBittorrent/blob/698f8dcb9cacf36a1e42ce7abfa7ac3d80df6476/src/gui/transferlistsortmodel.cpp#L72-L85

glassez replied:
> However, I did not convert an existing function into a template, because I believe that is already existing?:

Oh, right, I overlooked it (it's not very convenient to do all this through a smartphone browser).
Unfortunately, it is limited only to arithmetic types.
But it's not a problem to make a more generic template. You can delete all instances of `customCompare` except the one that accept `TagSet` as parameters, and add the following code instead:
```c++
    // consider negative values as invalid
    template <typename T>
        requires std::is_arithmetic_v<T>
    bool isValid(const T value)
    {
        return (value >= 0);
    }

    template <typename T>
    bool isValid(const T &value)
    {
        return value.isValid();
    }

    template <typename T>
    int customCompare(const T &left, const T &right)
    {
        const bool isLeftValid = isValid(left);
        const bool isRightValid = isValid(right);

        if (isLeftValid && isRightValid)
            return threeWayCompare(left, right);
        if (!isLeftValid && !isRightValid)
            return 0;
        return isLeftValid ? -1 : 1;
    }
```

ManiMatter replied:
Done - had to add a small refinement, since the 2nd template was also triggered for arithmetic types and the CI complained about template overloading. Hope the refinement makes sense?

[d486622](https://github.com/qbittorrent/qBittorrent/pull/20951/commits/d486622ad8a104657cb24aabc6226d272d7b41a6)

glassez replied:
> CI complained about template overloading.

Hmm... Clang compiled this without a problem.
`std::enable_if_t` looks cumbersome. I'd rather use a more modern approach with `if constexpr`:
```c++
    template <typename T>
    bool isValid(const T &value)
    {
        if constexpr (std::is_arithmetic_v<T>)
            return (value >= 0); // consider negative values as invalid
        else
            return value.isValid();
    }
```

glassez replied:
...or better add requirement to second template:
```c++
    template <typename T>
    concept Validateable = requires (T t) { {t.isValid()} -> std::same_as<bool>; };

    template <Validateable T>
    bool isValid(const T &value)
    {
        return value.isValid();
    }

    // consider negative values as invalid
    template <typename T>
        requires std::is_arithmetic_v<T>
    bool isValid(const T value)
    {
        return (value >= 0);
    }
```

ManiMatter replied:
updated ✔
https://github.com/qbittorrent/qBittorrent/pull/20951/commits/05c4f171d553e40dcdddf4b6db120531476134e4

ManiMatter replied:
done - https://github.com/qbittorrent/qBittorrent/pull/20951/commits/d8c743994449792b2ab1451ebc5a09d08ddb4ec4

glassez commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "transferlistsortmodel.h"
 
 #include <type_traits>
+#include <concepts>
```
```suggestion
#include <concepts>
#include <type_traits>
```

ManiMatter replied:
To keep them in alphabetical order?

Chocobo1 replied:
>To keep them in alphabetical order?

yes.

ManiMatter replied:
Happy to change that order. But what about the other inclusions? They are not in alphabetical order either.

> #include "transferlistsortmodel.h"
> 
> #include <concepts>
> #include <type_traits>
> 
> #include <QDateTime>
> 
> #include "base/bittorrent/infohash.h"
> #include "base/bittorrent/torrent.h"
> #include "transferlistmodel.h"

Just for my education and to understand how to do it "the right way" - Is there a reason for that specific order?

glassez replied:
> Just for my education and to understand how to do it "the right way" - Is there a reason for that specific order?

There are various reasons, some of which are objective, some of which are subjective - there are many discussions on this topic on the web. In general, we have drawn up our own rules based on them:
https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#7-header-inclusion-order

ManiMatter replied:
Thx for sharing, will study those

Chocobo1 commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -207,6 +213,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
     case TransferListModel::TR_SEEN_COMPLETE_DATE:
         return customCompare(leftValue.toDateTime(), rightValue.toDateTime());
 
+    case TransferListModel::TR_PRIVATE:
```
wrong order.

ManiMatter replied:
Could you elaborate please?

Chocobo1 replied:
Please keep them in alphabetical order.

ManiMatter replied:
done.

ManiMatter commented:
> > @glassez - I was under the impression that this one is pretty much done. Do you see remaining items that need to be addressed?
> 
> Yes. At least [#20951 (comment)](https://github.com/qbittorrent/qBittorrent/pull/20951#discussion_r1657005692).

Do you see any other items outstanding?

Chocobo1 commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -210,6 +216,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
     case TransferListModel::TR_DLLIMIT:
     case TransferListModel::TR_DLSPEED:
     case TransferListModel::TR_QUEUE_POSITION:
+    case TransferListModel::TR_PRIVATE:
```
```suggestion
    case TransferListModel::TR_PRIVATE:
    case TransferListModel::TR_QUEUE_POSITION:
```

glassez commented:
@ManiMatter 
Thank you!

stalkerok commented:
@ManiMatter, thanks!

stalkerok commented:
> Milestone
No milestone

Can this be added to 5.0?

glassez commented:
> > Milestone
> > No milestone
> 
> Can this be added to 5.0?

Did you test it carefully? Does it work as expected (display correct "privateness" value and "N/A" when torrent has no metadata received, sort correctly)?

stalkerok commented:
I tested this before the final changes and it worked fine. I'll test again a little later to say for sure.

ManiMatter commented:
> > Milestone
> > No milestone
> 
> Can this be added to 5.0?

Would it make sense that it goes together with https://github.com/qbittorrent/qBittorrent/pull/20833 and https://github.com/qbittorrent/qBittorrent/pull/20686, since they are somewhat related?

stalkerok commented:
There seems to be something wrong with the sorting
GUI
![2024-07-08_225829](https://github.com/qbittorrent/qBittorrent/assets/63958081/c898da76-a948-47f2-940f-b9f3d557640b) ![2024-07-08_230033](https://github.com/qbittorrent/qBittorrent/assets/63958081/10b2be0a-1147-4177-846b-06fed8a2135d)

WEB UI
![2024-07-08_230552](https://github.com/qbittorrent/qBittorrent/assets/63958081/fe740d91-880d-490f-b184-e1edd7ba018a) ![2024-07-08_230610](https://github.com/qbittorrent/qBittorrent/assets/63958081/7fe45dc0-fa57-4943-96a0-ee7d4c0613be)

glassez commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -209,6 +215,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
 
     case TransferListModel::TR_DLLIMIT:
     case TransferListModel::TR_DLSPEED:
+    case TransferListModel::TR_PRIVATE:
```
>There seems to be something wrong with the sorting

Damn it, we've overlooked it after all. This is the wrong place because it is eventually compared as an integer.
```c++
return customCompare(leftValue.toInt(), rightValue.toInt());
``` 
It should be moved outside of this group of cases and placed separately:
```c++
case TransferListModel::TR_PRIVATE:
    return customCompare(leftValue, rightValue);
```

glassez commented:
>There seems to be something wrong with the sorting

As for GUI:
https://github.com/qbittorrent/qBittorrent/pull/20951/files#r1669878145

As for WebUI:
The problem is most likely that it compares by string representation, and not by logical values, which I insisted on and which was eventually implemented (albeit with errors) in the GUI.

glassez commented:
@stalkerok 
Could you test #21039?

glassez commented:
> Seems the problem is more serious (provided `customCompare()` function is still unsuitable for `QVariant`).

#21041 should take care about it.

Chocobo1 commented on src/gui/transferlistsortmodel.cpp:
```diff
@@ -70,12 +61,27 @@ namespace
-        const bool isLeftValid = (left >= 0);
-        const bool isRightValid = (right >= 0);
+    template <typename T>
+    int customCompare(const T &left, const T &right)
+    {
+        const bool isLeftValid = isValid(left);
+        const bool isRightValid = isValid(right);
```
@glassez 
This new comparison logic isn't equivalent to the removed `int customCompare(const QDateTime &left, const QDateTime &right)`.
The `if (!isLeftValid && !isRightValid)` would produce unexpected results.

Furthermore, this `customCompare` was really intended for specific arithmetic types and not general use.
I would suggest reverting changes to this file in #21041.

glassez replied:
> Furthermore, this `customCompare` was really intended for specific arithmetic types and not general use.

I don't know how you define it, but this compare logic is really the same for all types for which we define the concept of "invalid value". The only difference is how the concept of "invalid value" is defined for a particular type.

glassez replied:
> I would suggest reverting these changes in #21041.

Done anyway. I don't want to waste my time on this issue anymore, it's already taken up too much of it.

Chocobo1 replied:
>I don't know how you define it, but this compare logic is really the same for all types for which we define the concept of "invalid value". The only difference is how the concept of "invalid value" is defined for a particular type.

For me, I would treat the comparison for each column as unique and it should be (easily) customizable. If some of them could share the same code then it is great and can be done. Also generalization should not become the final goal as it would hide the salient comparison of the column values (and possibly introduce slower comparison code).
